Add a prototype LSP

klarasm opened the pull request:
Prototype LSP. Mostly similar to the current plyghd prototype except some special cases. It now also gets the source file name from the LSP client.

klarasm committed:
add a simple lsp which listens on port 5007

klarasm committed:
send the diagram/accept messages as expected by KLighD

SamuelMiksits committed:
Add argument parsing for LSP and move main arguments module

Closes: klarasm/forsyde-devtools#1

klarasm committed:
start transforming ForSyDeIR to KGraph

klarasm committed:
get sourceUri from the LSP client

klarasm committed:
add basic outline of SKGraphSchema

klarasm committed:
remember the sourceUri

klarasm committed:
add colors

klarasm committed:
add KText

klarasm committed:
add KRoundedBendsPolyline

klarasm committed:
add "directed" edges

There should be a better way to do this but haven't gotten it to work yet.

klarasm committed:
add comments to SKGraphSchema

klarasm committed:
add comments to LSP

SamuelMiksits commented:
Can you give an example on how to run this (klighd-linux cli and forsyde-lsp commands) and what you expect it to do? I can't figure it out and just get some JSON rpc error but I am not sure how I am supposed to run it

klarasm commented:
I think it should be sufficient with:
```sh
cabal run forsyde-lsp-exe
# in another terminal
./klighd-cli --ls_port 5007 <path to a forsyde model>
```
If the browser is not opened you can go to the thing which klighd-cli outputs

SamuelMiksits commented on src/ArgumentsLSP.hs:
```diff
@@ -0,0 +1,79 @@
+    -- Formats
+    input :: Input
+  }
+
```
Ah my bad. It took me like an hour of debugging to implement `inputFile` where I tried different approaches, among them calling `fileName` from `inputFile`. But it ended up being the case that I didn't need this function. So can literally just

```diff
diff --git a/src/ArgumentsLSP.hs b/src/ArgumentsLSP.hs
index 75a2967..72f6ef4 100644
--- a/src/ArgumentsLSP.hs
+++ b/src/ArgumentsLSP.hs
@@ -26,12 +26,6 @@ data Arguments = Arguments
     input :: Input
   }
 
--- Function to create a filepath with file extension
-fileName :: ReadM (FilePath)
-fileName = str >>= returnFile
-  where
-    returnFile s = pure s
-
```

SamuelMiksits reviewed (approved):
Solid prototype, has some issues with rendering that I think we handled in Python that are not handled here yet, could be addressed with `case` in the future:
- Edge port should not have a token rate when:
  - Source is global input
  - Destination is global output
  - Source is delay
  - Destionation is delay
- Edge port should not have a "arrow" when destination is delay
- Size constraint is weird for delay, so usually ends up being an ellipsoid rather than a circle. But we were considering if we should instead add delay as an "ascii ball" edge label with inline label setting which would fix the other delay issues that I mentioned. But could also interfere with our edge labels (like "s_1" etc)? I forget how it is implemented, you might be able to just put the inline label constraint on that label on top of having the "s_1" label that acts normally. 

<img width="1897" height="977" alt="sdf_002" src="https://github.com/user-attachments/assets/9e77a90e-d2ed-4ce6-9b04-11af5372d37b" />

klarasm committed:
do not show rates for delay and global input/output

also do not make a directed edge to the delay.

klarasm committed:
add signal labels

klarasm commented:
> Solid prototype, has some issues with rendering that I think we handled in Python that are not handled here yet, could be addressed with `case` in the future:
> 
>    * Edge port should not have a token rate when:
>      
>      * Source is global input
>      * Destination is global output
>      * Source is delay
>      * Destionation is delay
>
>    * Edge port should not have a "arrow" when destination is delay
>
>    * Size constraint is weird for delay, so usually ends up being an ellipsoid rather than a circle. But we were considering if we should instead add delay as an "ascii ball" edge label with inline label setting which would fix the other delay issues that I mentioned. But could also interfere with our edge labels (like "s_1" etc)? I forget how it is implemented, you might be able to just put the inline label constraint on that label on top of having the "s_1" label that acts normally.

@SamuelMiksits:

I think I resolved most of these now. I kept the delay as an ellipse but removed some size constraints, made it smaller, added background color, and removed the label. I also added the signal labels.

![SDF_example_002(1)](https://github.com/user-attachments/assets/67a7d40e-8053-41b4-8142-7bb916a0420e)

klarasm committed:
simplify getFilePathFromClient

klarasm commented:
I'll merge this now so other people can work on it more easily.
